### ISSUE TITLE
Add psalm.xml option to specify additional include paths when analyzing require_once, etc?

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#268]
This will reduce false positives analyzing projects which use relative include paths in some places.

`stream_resolve_include_path($relative_path_from_codebase)` may help in writing an implementation (and temporarily overriding set_include_path()).

Example syntax (Naming could definitely be better):

```xml
<!-- set_include_path('.' . ':' . get_include_path()); -->
<add_current_directory_to_include_path />
<!-- Relative paths to the root project directory, possible even absolute paths -->
<root_include_path>path/to/manually_added/subfolder</root_include_path>
<root_include_path>path/to/other/manually_added/subfolder</root_include_path>
```

The code in question where changes would be made

```php
// src/Psalm/Checker/StatementsChecker.php

            $reduce_pattern = '/\/[^\/]+\/\.\.\//';

            while (preg_match($reduce_pattern, $path_to_file)) {
                $path_to_file = preg_replace($reduce_pattern, DIRECTORY_SEPARATOR, $path_to_file);
            }

            // if the file is already included, we can't check much more
            if (in_array($path_to_file, get_included_files(), true)) {
                return null;
            }
```

---

An alternative would be for psalm to try to parse set_include_path() calls to see any project/library directories that are added. This wouldn't pick up php.ini settings for include_path that are different from the one used by PHP binary running psalm.

- This looks possible but impractical (would need to check for PATH_SEPARATOR or ':' when parsing AST argument of set_include_path, and defer emitting issues)

**Comment by TysonAndre** [comment]
Related bug:

I have foo/bar/b/c.php and foo/www/php-resource in the root directory of a project

c.php has `include('../../www/php-resource')` as a statement. That would be `'foo/bar/b/../../www/php-resource'`, which exists, because `.` (directory of the current PHP file) is in the include path.

Expected: No error
Observed: MissingFile for that `include()`

**Comment by muglug** [comment]
I think I've improved handling of included files sufficiently (to support Wordpress analysis). I've filed the include path issue here: https://github.com/vimeo/psalm/issues/891
